pkg/agent/replica_set_port_manager.go

```diff
@@ -0,0 +1,108 @@
+package agent
+
+import (
+	"github.com/mongodb/mongodb-kubernetes-operator/pkg/automationconfig"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type ReplicaSetPortManager struct {
+	log              *zap.SugaredLogger
+	expectedPort     int
+	currentPodStates []PodState
+	currentAC        automationconfig.AutomationConfig
+}
+
+func NewReplicaSetPortManager(log *zap.SugaredLogger, expectedPort int, currentPodStates []PodState, currentAC automationconfig.AutomationConfig) *ReplicaSetPortManager {
+	return &ReplicaSetPortManager{log: log, expectedPort: expectedPort, currentPodStates: currentPodStates, currentAC: currentAC}
+}
+
+func (r *ReplicaSetPortManager) GetPortsModification() automationconfig.Modification {
+	portMap, _, _ := r.calculateExpectedPorts()
+	r.log.Debugf("Calculated process port map: %+v", portMap)
+	return func(config *automationconfig.AutomationConfig) {
+		for i := range config.Processes {
+			process := config.Processes[i]
+			process.SetPort(portMap[process.Name])
+		}
+	}
+}
+
+// GetServicePorts returns array of corev1.ServicePort.
+// If there is no port change in progress it returns expectedPort named "mongodb".
+// If there is port change in progress then it returns both ports: old named "mongodb" and new named "mongodb-new".
+// When port change is finished, it falls back to the first case (no port change) and "mongodb-new" will be renamed to "mongodb".
+func (r *ReplicaSetPortManager) GetServicePorts() []corev1.ServicePort {
+	_, portChangeRequired, oldPort := r.calculateExpectedPorts()
+
+	if !portChangeRequired || oldPort == r.expectedPort {
+		return []corev1.ServicePort{{
+			Port: int32(r.expectedPort),
+			Name: "mongodb",
+		}}
+	}
+
+	servicePorts := []corev1.ServicePort{
+		{
+			Port: int32(r.expectedPort),
+			Name: "mongodb-new",
+		},
+		{
+			Port: int32(oldPort),
+			Name: "mongodb",
+		},
+	}
+
+	r.log.Debugf("Port change in progress, setting service ports: %+v", servicePorts)
+
+	return servicePorts
+}
+
+func (r *ReplicaSetPortManager) calculateExpectedPorts() (map[string]int, bool, int) {
```

Comment:
This method needs a comment as to what `calculateExpectedPorts` means.

Reply:
Done. Also added some more docs also in code as the process is not straightforward at first glance.